src/finch/tensor.py

```diff
@@ -1,18 +1,179 @@
+            )
+
+    @property
+    def dtype(self) -> jc.TypeValue:
+        return jl.eltype(self._obj)
+
+    @property
+    def ndim(self) -> int:
+        return jl.ndims(self._obj)
+
+    @property
+    def shape(self) -> Tuple[int, ...]:
+        return jl.size(self._obj)
+
+    @property
+    def size(self) -> int:
+        return np.prod(self.shape)
+
     def __pos__(self):
-        return Tensor(jl.Base.broadcast(jl.seval("+"), self.data))
-    def __neg__(self):    
-        return Tensor(jl.Base.broadcast(jl.seval("-"), self.data))
+        return Tensor(jl_data=jl.Base.broadcast(jl.seval("+"), self._obj))
+
     def __add__(self, other):
-        return Tensor(jl.Base.broadcast(jl.seval("+"), self.data, other.data))
+        return Tensor(jl_data=jl.Base.broadcast(jl.seval("+"), self._obj, other._obj))
+
+    def __mul__(self, other):
+        return Tensor(jl_data=jl.Base.broadcast(jl.seval("*"), self._obj, other._obj))
+
+    def __sub__(self, other):
+        return Tensor(jl_data=jl.Base.broadcast(jl.seval("-"), self._obj, other._obj))
+
+    def __truediv__(self, other):
+        return Tensor(jl_data=jl.Base.broadcast(jl.seval("/"), self._obj, other._obj))
+
+    def todense(self) -> np.ndarray:
+        shape = jl.size(self._obj)
+
+        # create fully dense array and access `val`
+        dense_lvls = jl.Element(jl.default(self._obj))
+        for _ in shape:
+            dense_lvls = jl.Dense(dense_lvls)
+        dense_tensor = jl.Tensor(dense_lvls, self._obj).lvl
+        for _ in shape:
+            dense_tensor = dense_tensor.lvl
+
+        return np.array(dense_tensor.val).reshape(shape, order="F")
+
+
+class COO(Tensor):
+    def __init__(self, coords, data, shape, fill_value=0.0):
+        assert coords.ndim == 2
+        ndim = len(shape)
+
+        lvl = jl.Element(fill_value, jl.Vector(data))
+        ptr = jl.Vector[jl.Int]([1, len(data) + 1])
+        tbl = tuple(jl.Vector(coords[i, :] + 1) for i in range(ndim))
+
+        jl_data = jl.SparseCOO[ndim](lvl, shape, ptr, tbl)
+
+        self._obj = jl.Tensor(jl_data)
+
+
+class _Compressed2D(Tensor):
+    def __init__(self, arg, shape, fill_value=0.0):
+        assert isinstance(arg, tuple) and len(arg) == 3
+        assert len(shape) == 2
+
+        data, indices, indptr = arg
+        data = jl.Vector(data)
+        indices = jl.Vector(indices + 1)
+        indptr = jl.Vector(indptr + 1)
+
+        lvl = jl.Element(fill_value, data)
+        jl_data = self.get_jl_data(shape, lvl, indptr, indices)
+        self._obj = jl.Tensor(jl_data)
+
+    @abstractmethod
+    def get_jl_data(
+        self,
+        shape: Tuple[int, int],
+        lvl: jc.AnyValue,
+        indptr: jc.VectorValue,
+        indices: jc.VectorValue,
+    ) -> jc.AnyValue:
+        ...
+
+
+class CSC(_Compressed2D):
+    def get_jl_data(self, shape, lvl, indptr, indices):
+        return jl.Dense(jl.SparseList(lvl, shape[0], indptr, indices), shape[1])
+
+
+class CSR(_Compressed2D):
```

Comment:
This is not CSR, unfortunately. You should instead create `jl.swizzle(jl.Dense(jl.SparseList(lvl, shape[0], indptr, indices), shape[1]), [2, 1])`

Reply:
Ah, right!
I've got a question about `swizzle` - I understand that this operation creates a new Tensor-like object: How should I convert it to an actual Tensor?

```python
from src import finch
import numpy as np

jl = finch.jl  # to just operate on julia directly
A = np.array([[0,0,4], [1,0,0], [2,0,5], [3,0,0]])

A_sparse = jl.Tensor(jl.Dense(jl.SparseList(jl.Element(0.0))), A)
A_sparse + A_sparse  # works

result = jl.swizzle(A_sparse.lvl, 2, 1)
result + result  # it isn't a Tensor
```

Reply:
Hmm. Currently the mode ordering is handled with a separate object, the SwizzledArray. Perhaps the approach here would be to mirror that for now, using a SwizzledArray Object. By default, we should probably reverse row to column major in the wrapper as well, so Finch tensors should be row major by default in python.

Reply:
I searched Finch.jl codebase and I found that SwizzleArray is converted to Tensor in one of the extensions:

https://github.com/willow-ahrens/Finch.jl/blob/1c7831df99dfcad763e6f9e90c18d9163833b514/ext/SparseArraysExt.jl#L29-L32

I tried to implement `CSR` the same way:
```python
class CSR(_Compressed2D):
    def get_jl_data(self, shape, lvl, indptr, indices):
        a = jl.swizzle(jl.Dense(jl.SparseList(lvl, shape[0], indptr, indices), shape[1]), 2, 1)
        return jl.Tensor(jl.Dense(jl.SparseList(jl.Element(0.0))), result)
```
but the last line (taken from the extension) fails with:
```
JuliaError: MethodError: no method matching ndims(::Type{DenseLevel{Int64, SparseListLevel{Int64, Vector{Int64}, Vector{Int64}, ElementLevel{0.0, Float64, Int64, Vector{Float64}}}}})
```